### PR TITLE
Allow structs and enums to be imported from elsewhere within the crate

### LINKED ISSUE #294: [Bug] `frb_codegen` fails with enum as a function argument, because only types defined in the input file are transformed
### Describe the bug

`frb_codegen` fails when an enum is given as a function argument.

### To Reproduce

1. Start with the pure dart example
2. Add the following to `rust/src/data.rs`:
```rust
pub enum MyEnum {
    One,
    Two,
}
```
3. Add the following to `rust/src/lib.rs`:
```rust
mod data;
```
4. Add the following to `rust/src/api.rs`:
```rust
use crate::data::MyEnum;

pub fn do_something(data: MyEnum) -> Result<i32> {
    match data {
        MyEnum::One => Ok(1),
        MyEnum::Two => Ok(2),
    }
}
```
5. Run codegen:
```ps1
flutter_rust_bridge_codegen --rust-input rust/src/api.rs --dart-output 
dart/lib/bridge_generated.dart
```

### Expected behavior

Codegen succeeds.

### Codegen logs

```shell
[2022-01-05T23:24:28Z INFO  flutter_rust_bridge_codegen] Picked config: Opts { rust_input_path: "C:\\Users\\qbgee\\Documents\\Code\\flutter_rust_bridge\\frb_example\\pure_dart\\rust/src/api.rs", dart_output_path: "C:\\Users\\qbgee\\Documents\\Code\\flutter_rust_bridge\\frb_example\\pure_dart\\dart/lib/bridge_generated.dart", c_output_path: "C:\\Users\\qbgee\\AppData\\Local\\Temp\\.tmp9PsTF6.h", rust_crate_dir: "C:\\Users\\qbgee\\Documents\\Code\\flutter_rust_bridge\\frb_example\\pure_dart\\rust", rust_output_path: "C:\\Users\\qbgee\\Documents\\Code\\flutter_rust_bridge\\frb_example\\pure_dart\\rust/src\\bridge_generated.rs", class_name: "FlutterRustBridgeExample", dart_format_line_length: 80, skip_add_mod_to_lib: false, llvm_path: "", llvm_compiler_opts: "" }
[2022-01-05T23:24:28Z INFO  flutter_rust_bridge_codegen] Phase: Parse source code to AST
[2022-01-05T23:24:28Z INFO  flutter_rust_bridge_codegen] Phase: Parse AST to IR
[2022-01-05T23:24:28Z DEBUG flutter_rust_bridge_codegen::parser] parse_function function name: Ident(do_something)
[2022-01-05T23:24:28Z DEBUG flutter_rust_bridge_codegen::parser] parse_type: MyEnum
thread 'main' panicked at 'parse_type failed for ty=MyEnum', C:\Users\qbgee\.cargo\registry\src\github.com-1ecc6299db9ec823\flutter_rust_bridge_codegen-1.13.0\src\parser.rs:151:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### Generated binding code

```shell
N/A, since codegen failed
```

### OS

Windows 10 21H1

### Version of `flutter_rust_bridge_codegen`

1.13.0

### Flutter info

```shell
[√] Flutter (Channel stable, 2.8.1, on Microsoft Windows [Version 10.0.19043.1415], locale en-US)
    • Flutter version 2.8.1 at C:\Users\qbgee\Documents\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (3 weeks ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[X] Android toolchain - develop for Android devices
    X Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Build Tools 2019 16.10.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools
    • Visual Studio Build Tools 2019 version 16.10.31424.327
    • Windows 10 SDK version 10.0.18362.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[√] VS Code (version 1.63.2)
    • VS Code at C:\Users\qbgee\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.29.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19043.1415]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 96.0.4664.110
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 96.0.1054.62

! Doctor found issues in 2 categories.
```

### Version of `clang++`

I don't think I have `clang++` installed, though I may be wrong.

### Version of `cbindgen`

0.20.0

### Version of `ffigen`

No `ffigen` listed in `pubspec.yaml` for the pure dart example.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by SecondFlight**
Resolves #294.

This PR adds a few things:
- A `source_graph` module, descried [here](https://github.com/fzyzcjy/flutter_rust_bridge/issues/249#issuecomment-1026391897)
- A few modifications to codegen:
  - Users must now supply a `--cargo-path` argument which indicates the `Cargo.toml` location for their crate
  - Struct and enum maps now collect structs and enums from across the crate
  - Relevant structs and enums are imported into the generated rust file

And it has a few potential issues:
- Except for `source_graph.rs`, this implementation is a bit hacky. Here are the problems that I see, though there are likely more:
  - `source_graph.rs` understands Rust modules and imports, but the code that uses it does not.
  - As a direct example of the above, this PR still has `flutter_rust_bridge` collect structs and enums into a struct map and an enum map. This means that a crate cannot have multiple structs or enums in different scopes with the same name.
- I wasn't very careful with data cloning. It's possible that I'm doing it in places that I don't need to.

I think some non-trivial refactoring will be needed to support #249. `source_graph.rs` still has some oversights, and I think `parser.rs` takes too many shortcuts to do large-scale module parsing without some work.

### CODE REVIEW

**Comment by fzyzcjy** [comment]
Excellent job!

**Comment by SecondFlight** [comment]
Thanks! I'll un-draft this once I appease CI, but otherwise this should be ready for review. Please don't hesitate to tear it apart. I'm not prepared to do the refactor described above, but I also don't want to leave a mess for anyone else.

File: frb_codegen/src/main.rs [hunk 1]
```diff
@@ -19,11 +19,12 @@ mod generator_dart;
 mod generator_rust;
 mod others;
 mod parser;
+mod source_graph;
 mod transformer;
 mod utils;
 
 fn main() {
-    env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();
+    env_logger::Builder::from_env(Env::default().default_filter_or("debug")).init();
```
Comment (fzyzcjy) [inline frb_codegen/src/main.rs:27]: What about keeping the default filter INFO, o/w users will see lots of log. And you can always change the type via env variables when you are debugging

**Comment by SecondFlight** [comment]
Oh, so this is a weird problem:
https://github.com/fzyzcjy/flutter_rust_bridge/runs/5102997820?check_suite_focus=true

`cargo fmt` is failing due to an internal error, both locally for me and in CI. I'll dig into this some and see what I can find. It started when I added some `panic!` statements in `source_graph.rs`.

File: frb_codegen/src/main.rs (hunk 1 above)
Comment (SecondFlight) [reply frb_codegen/src/main.rs:27]: Ah, just an oversight. Changing that now.

File: frb_codegen/src/config.rs [hunk 2]
```diff
@@ -12,6 +12,9 @@ use toml::Value;
 #[derive(StructOpt, Debug, PartialEq, Deserialize)]
 #[structopt(setting(AppSettings::DeriveDisplayOrder))]
 pub struct RawOpts {
+    /// Path of input cargo.toml file
+    #[structopt(short, long)]
+    pub cargo_path: String,
```
Comment (Desdaemon) [inline frb_codegen/src/config.rs:17]: You might have to remove the short flag here since it's conflicting with c_output.

**Comment by SecondFlight** [comment]
Re: rustfmt issue, I think I've isolated the root cause. I submitted an issue to `rustfmt` here: https://github.com/rust-lang/rustfmt/issues/5226

File: frb_codegen/src/generator_rust.rs [hunk 3]
```diff
@@ -36,6 +37,26 @@ impl Generator {
...
+            .iter()
+            .map(|api_type| self.generate_import(api_type, api_file))
+            .collect();
+        let imports: Vec<String> = input_type_imports
+            .into_iter()
+            .chain(output_type_imports.into_iter())
+            .filter(|import| import.is_some())
+            .map(|import| import.unwrap())
+            // Don't include imports from the API file
+            .filter(|import| !import.starts_with(&format!("use crate::{}::", rust_wire_mod)))
```
[the remark this answers is not in the document]
Comment (SecondFlight) [reply frb_codegen/src/generator_rust.rs:51]: Done, though `clippy` decided this should be simply `.flatten()`.

**Comment by SecondFlight** [comment]
Looks like I'm getting a deprecation warning on a file I didn't modify (see [here](https://github.com/fzyzcjy/flutter_rust_bridge/runs/5119000229?check_suite_focus=true)).

**Comment by SecondFlight** [comment]
I think I've solved all the build issues I can. Let me know if there's anything else I can address.

Re: formatting issue, it appears to be somewhat-expected (if maybe not optimal) behavior. See [here](https://github.com/rust-lang/rustfmt/issues/5170#issuecomment-1016630482) for an explanation.

**Comment by fzyzcjy** [comment]
> Looks like I'm getting a deprecation warning on a file I didn't modify (see [here](https://github.com/fzyzcjy/flutter_rust_bridge/runs/5119000229?check_suite_focus=true)).

Maybe suppress it (since that is only a sample code, not core code)

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -474,26 +474,26 @@ jobs:
       - name: Run codegen to pure_dart example (non-windows)
         if: ${{ matrix.os.family != 'windows' }}
         working-directory: ./frb_codegen
-        run: cargo run --package flutter_rust_bridge_codegen --bin flutter_rust_bridge_codegen -- --rust-input ../frb_example/pure_dart/rust/src/api.rs --dart-output ../frb_example/pure_dart/dart/lib/bridge_generated.dart --dart-format-line-length 120
+        run: cargo run --package flutter_rust_bridge_codegen --bin flutter_rust_bridge_codegen -- --cargo-path ../frb_example/pure_dart/rust/Cargo.toml --rust-input ../frb_example/pure_dart/rust/src/api.rs --dart-output ../frb_example/pure_dart/dart/lib/bridge_generated.dart --dart-format-line-length 120
```
Comment (fzyzcjy) [inline .github/workflows/ci.yaml:477]: Is it a mandatory flag? If so the developer experience seems to decrease.

Maybe we can infer this location, by looking at `--rust-input`'s nearest parent directories which contains a `Cargo.toml` file? And we can keep this flag as a fallback when the user really wants to specify it manualy.

Comment (Desdaemon) [reply .github/workflows/ci.yaml:477]: Some bike-shedding, but this flag can be better named manifest-path to fit with other Rust tools.

**Comment by fzyzcjy** [comment]
Btw I should move you up in the all-contributors section of README after the PR is merged (since I do not want to create code conflicts). Remind me if I forget ;)

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (SecondFlight) [reply .github/workflows/ci.yaml]: Turns out I didn't actually read the code there. The root rust folder is already inferred (with an optional arg override), and the manifest path can be trivially inferred from that. I've eliminated the argument altogether as I think being able to override the root folder is good enough.

**Comment by SecondFlight** [comment]
Huh, I'm not sure what I did to make the Windows build angry, especially since I'm developing on Windows. Aside from that, this PR should be good to merge.

**Comment by SecondFlight** [comment]
Looking a bit more closely at that, I don't think that failure is related to this PR. It started after [this one-line change](https://github.com/fzyzcjy/flutter_rust_bridge/pull/319/commits/5c7616a0b2a64c63d52607cb8ab60e77e1041369), which adds a warning ignore in the example that's not failing. The build before was successful 😕

**Comment by fzyzcjy** [comment]
Well, then ignore this strange window failure and let us address it later in another pr maybe.

I have checked android PR failure, seems also unrelated to this pr.

Thus, LGTM.

@Desdaemon what do you think?

**Comment by Desdaemon** [comment]
I can't check the Windows failures from my phone, but given that other tests run fine, LGTM.

**Comment by fzyzcjy** [comment]
🎉

**Comment by fzyzcjy** [comment]
- [ ] move @SecondFlight in all-contributors
- [ ] modify the book

**Pull request merged**